File: go/src/internal/htmlrewrite/srcset.go
package htmlrewrite

import "strings"

// rewriteSrcset rewrites every URL in an HTML srcset attribute, leaving the
// width/density descriptor (`1x`, `640w`, etc.) attached to its original URL.
//
// Spec: https://html.spec.whatwg.org/multipage/images.html#parsing-a-srcset-attribute
//
// Per spec, URLs end at ASCII whitespace — commas are valid URL characters and
// appear in Cloudflare Image Resizing paths like:
//
//	/cdn-cgi/image/width=640,quality=75,format=auto/https://origin.com/img.jpg
//
// Candidate separators are commas that appear AFTER a descriptor (or after the
// URL token when no descriptor is present). We detect them by reading the URL
// token up to the first whitespace, then reading the descriptor up to the next
// comma.
func rewriteSrcset(srcset string, rewriteURL func(string) string) string {
	var out []string
	s := srcset

	for {
		// Per spec step 4: skip leading whitespace and commas (separators from
		// the previous candidate, or leading junk).
		s = strings.TrimLeft(s, " \t\n\r\v,")
		if s == "" {
			break
		}

		// Per spec step 6: URL = chars until first ASCII whitespace.
		wsIdx := strings.IndexAny(s, " \t\n\r\v")
		var urlToken string
		if wsIdx < 0 {
			urlToken = s
			s = ""
		} else {
			urlToken = s[:wsIdx]
			s = s[wsIdx:]
		}

		// Per spec step 8: trailing commas on the URL itself are separators.
		urlToken = strings.TrimRight(urlToken, ",")
		if urlToken == "" {
			continue
		}

		// Skip whitespace between URL and optional descriptor.
		s = strings.TrimLeft(s, " \t\n\r\v")

		// Descriptor: chars until the next comma (the next candidate separator).
		commaIdx := strings.Index(s, ",")
		var descriptor string
		if commaIdx < 0 {
			descriptor = strings.TrimSpace(s)
			s = ""
		} else {
			descriptor = strings.TrimSpace(s[:commaIdx])
			s = s[commaIdx:] // leave the comma for the next iteration's TrimLeft
		}

		if descriptor != "" {
			out = append(out, rewriteURL(urlToken)+" "+descriptor)
		} else {
			out = append(out, rewriteURL(urlToken))
		}
	}

	return strings.Join(out, ", ")
}

File: ts/client/src/patches/srcset.ts
// Spec-compliant srcset rewriter.
//
// Per https://html.spec.whatwg.org/multipage/images.html#parse-a-srcset-attribute
// URLs end at ASCII whitespace — commas within a URL (e.g. Cloudflare Image
// Resizing paths like /cdn-cgi/image/width=640,quality=75,format=auto/...) are
// valid and must NOT be treated as candidate separators. The separator between
// candidates is a comma that appears AFTER a descriptor (or after a URL-only
// entry with no descriptor).
//
// Algorithm mirrors go/src/internal/htmlrewrite/srcset.go — keep in sync.

const WS = /[ \t\n\r\v]/;

export function rewriteSrcsetAttribute(
    srcset: string,
    rewriteOne: (url: string) => string,
): string {
    const out: string[] = [];
    let s = srcset;

    while (true) {
        // Skip leading whitespace and commas (separators from previous candidate).
        s = s.replace(/^[ \t\n\r\v,]+/, "");
        if (s === "") break;

        // URL token ends at first ASCII whitespace.
        const wsIdx = s.search(WS);
        let urlToken: string;
        if (wsIdx === -1) {
            urlToken = s;
            s = "";
        } else {
            urlToken = s.slice(0, wsIdx);
            s = s.slice(wsIdx);
        }

        // Trailing commas on the URL itself are candidate separators.
        urlToken = urlToken.replace(/,+$/, "");
        if (urlToken === "") continue;

        // Skip whitespace between URL and optional descriptor.
        s = s.replace(/^[ \t\n\r\v]+/, "");

        // Descriptor: everything up to the next comma.
        const commaIdx = s.indexOf(",");
        let descriptor: string;
        if (commaIdx === -1) {
            descriptor = s.trim();
            s = "";
        } else {
            descriptor = s.slice(0, commaIdx).trim();
            s = s.slice(commaIdx); // leave the comma for next iteration's trim
        }

        if (descriptor !== "") {
            out.push(`${rewriteOne(urlToken)} ${descriptor}`);
        } else {
            out.push(rewriteOne(urlToken));
        }
    }

    return out.join(", ");
}
